# Phase 1: Data Understanding

In [ ]:
## Importing pandas
import pandas as pd

## Loading table No.1 - Orders
orders = pd.read_csv("olist_orders_dataset.csv")

## Looking at the raw data and finding the PK
orders.head()
print(orders.columns)
print("The number of rows is :", orders.shape[0])
print("The number of unique order IDs is :", orders['order_id'].nunique())

## Looking for missing values
print("Missing values:\n", orders.isnull().sum())

orders.head()

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='object')
The number of rows is : 99441
The number of unique order IDs is : 99441
Missing values:
 order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


### Orders table conclusions
In this table each row represents a different order.
1. **Primary key**: `order_id`.
2. **Foreign key**: `customer_id` (to the customers table).
3. **Interesting points**: The time difference between the `order_estimate_delivery_date` and the `order_delivered_customer_date`.
4. **Missing values**: There are missing values in three columns: `order_approved_at`, `order_delivered_customer_date` and `order_delivered_customer_date`.  


In [ ]:
## Loading table No.2 - customers
customers = pd.read_csv("olist_customers_dataset.csv")

## Looking into the raw data and finding the PK
customers.head()
print(customers.columns)
print("The number of rows is :", customers.shape[0])
print("The number of unique customer IDs is :", customers['customer_id'].nunique()) # PK
print("The number of unique customer unique IDs is :", customers['customer_unique_id'].nunique())

## Looking for missing values
print("Missing values:\n", customers.isnull().sum())

Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='object')
The number of rows is : 99441
The number of unique customer IDs is : 99441
The number of unique customer unique IDs is : 96096
Missing values:
 customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64


### Customers table conclusions
In this table each row represents a customer in a specific order.
1. **Primary key**: `customer_id`. The number of unique `customer_id` values equals the number of rows,
while the number of `customer_unique_id` values is lower. This indicates that some customers placed multiple orders. For customer-level analysis, `customer_unique_id` will therefore be used
2. **Interesting points**: As mentioned, we can see that some of the customers placed more than one order.
`customer_zip_code_prefix`, `customer_city`,`customer_state` can be referenced in the geolocation table. There may be a connection between the geolocation of the seller and the freight prices.
3. There are no missing values in this table.

In [ ]:
## Loading table No.3 - order_items
order_items = pd.read_csv("olist_order_items_dataset.csv")

## Looking into the raw data and finding the PK
order_items.head()
print(order_items.columns)
print("The number of rows is :", order_items.shape[0])
print("The number of unique order item IDs is :", order_items['order_item_id'].nunique()) # Less than the total rows
print("The number of unique order IDs is :", order_items['order_id'].nunique()) # Less than the total rows

## Finding the composite key of the table
print("Combined fileds: ", order_items.duplicated(subset = ['order_item_id', 'order_id']).sum())

## Looking for missing values
print("Missing values:\n", order_items.isnull().sum())

order_items.head()


Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='object')
The number of rows is : 112650
The number of unique order item IDs is : 21
The number of unique order IDs is : 98666
Combined fileds:  0
Missing values:
 order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


### Order items table conclusions
In this row each row represents an item in a specific order.
 1. **Cpmposite key**: `order_id` is not unique because a single order may contain multiple items. The combination of `order_id` and `order_item_id` was therefore tested as a potential composite key, and was found as a valid composite key.
 2. **Foreign keys**: `product_id` (to the product table) and `seller_id` (to the seller table).
 3. **Interesting points**: What influences the freight price. There are orders with more than one item.
 4. There are no missing values in the table.



In [ ]:
from numpy._core.fromnumeric import prod

## Loading table No.4 - products
products = pd.read_csv("olist_products_dataset.csv")

## Looking at the data and finding PK
products.head()
print(products.columns)
print("The number of rows is :", products.shape[0])
print("The number of unique product IDs is :", products['product_id'].nunique()) # PK

## Looking for missing values
print("Missing values:\n", products.isnull().sum())

## Checking whether the missing values are of the same rows
missing_info = products[
    ['product_category_name',
     'product_name_lenght',
     'product_description_lenght',
     'product_photos_qty']
].isnull().all(axis = 1).sum()

missing_dimensions = products[
    ['product_weight_g',
     'product_length_cm',
     'product_height_cm',
     'product_width_cm']
].isnull().all(axis = 1).sum()

print((missing_info & missing_dimensions).sum())

Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='object')
The number of rows is : 32951
The number of unique product IDs is : 32951
Missing values:
 product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64
2


### Product table conclusions
In this table, each row represents a product and its dimensions.
1. **Primary key**: `product_id`.
2. **Foreign key**: `Product_category_name` (to the table product_category_name_translation).
3. **Interesting points**: The dimensions of a product may impact the price of the fright or the shipping.
4. Missing values show a consistent pattern across the products dataset:

- 610 products are missing category and listing information, including product name length, description length, and photo quantity.
- Of these 610 products, 2 are also missing all physical product attributes: weight, length, height, and width.

No records were removed at this stage. Missing values will be handled according to the requirements of each analysis.

In [ ]:
## Loading table No.5 - product_category_name_translation
translation = pd.read_csv("product_category_name_translation.csv")

## Looking into the raw data and finding the PK
translation.head()
print(translation.columns)
print("The number of rows is :", translation.shape[0])
print("The number of unique product category name is :", translation['product_category_name'].nunique()) # PK

## Looking for missing values
print("Missing values:\n", translation.isnull().sum())

Index(['product_category_name', 'product_category_name_english'], dtype='object')
The number of rows is : 71
The number of unique product category name is : 71
Missing values:
 product_category_name            0
product_category_name_english    0
dtype: int64


### Translation table conclusions
In this table in each row there is a translation of the product category name from Portuguese to English.
1. **Primary key**: both columns can be a primary key. I'll go with `product_category_name`.
2. **Interesting points**: In order to work with the product table we will have to join the tables to get the name of the categories in English.
3. There are no missing values in this table.

In [ ]:
## Loading table No. 6 - sellers
sellers = pd.read_csv("olist_sellers_dataset.csv")

## Looking into the raw data and finding the PK
sellers.head()
print(sellers.columns)
print("The number of rows is :", sellers.shape[0])
print("The number of unique seller IDs is :",sellers['seller_id'].nunique()) # PK

## Looking for missing values
print("Missing values:\n", sellers.isnull().sum())

Index(['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state'], dtype='object')
The number of rows is : 3095
The number of unique seller IDs is : 3095
Missing values:
 seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64


,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


### Sellers table conclusions
In this table each row represents a different seller.
1. **Primary key**: `seller_id`.
2. **Interesting points**: `seller_zip_code_prefix`, `seller_city`,`seller_state` can be referenced in the geolocation table. There may be a connection between the geolocation of the seller and the freight prices.
3. There are no missing values in this table.

In [ ]:
## Loading table No.7 - payments
payments = pd.read_csv("olist_order_payments_dataset.csv")

## Looking into the raw data and finding the PK
payments.head()
print(payments.columns)
print("The number of rows is :", payments.shape[0])
print("The number of unique order_IDs is :" ,payments['order_id'].nunique()) # Less than the number of rows
print("The number of unique payment sequential is :", payments['payment_sequential'].nunique()) # Less than the number of rows

## Finding the composite key of the table
print("Combined fileds: ", payments.duplicated(subset = ['order_id', 'payment_sequential']).sum())

## Looking for missing values
print("Missing values:\n", payments.isnull().sum())

Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='object')
The number of rows is : 103886
The number of unique order_IDs is : 99440
The number of unique payment sequential is : 29
Combined fileds:  0
Missing values:
 order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64


### Payments table conclusions
In this table each row represents a payment within a specific order.
1. **Composite key**: Since both `order_id` and `payment_sequential` are not unique, there will be a composite key of the two of the fileds.
2. There are no missing values in this table.

In [ ]:
## Loading table No.8 - reviews
reviews = pd.read_csv("olist_order_reviews_dataset.csv")

## Looking into the raw data and finding the PK
reviews.head()
print(reviews.columns)
print("The number of rows is :", reviews.shape[0])
print("The number of unique review IDs is :", reviews['review_id'].nunique()) # Less than the number of rows
print("The number of unique order IDs is :", reviews['order_id'].nunique()) # Less than the number of rows

## Finding the composite key of the table
print("Combined fileds: ", reviews.duplicated(subset = ['order_id', 'review_id']).sum()) # PK

## Understanding what are the values in this column
print(reviews['review_score'].unique())

## Finding the number of recourring ID's
review_id_counts = reviews['review_id'].value_counts()
order_id_counts = reviews['order_id'].value_counts()

print("Review IDs appearing more than once:",
      (review_id_counts > 1).sum())

print("Order IDs appearing more than once:",
      (order_id_counts > 1).sum())

## Looking into missing values
print("Missing values:\n", reviews.isnull().sum())

Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='object')
The number of rows is : 99224
The number of unique review IDs is : 98410
The number of unique order IDs is : 98673
Combined fileds:  0
[4 5 1 3 2]
Review IDs appearing more than once: 789
Order IDs appearing more than once: 547
Missing values:
 review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64


### Reviews table Conclusions
In this table each row represents a specific review in a specific order.
1. **Composite key**:
Neither review_id nor order_id is unique in the reviews dataset. The combination of `review_id` and `order_id` uniquely identifies each record.
2. **Interesting points**: Duplicate review_id values across different orders contain identical review information, indicating that the same review record can be associated with multiple orders. In contrast, orders associated with multiple review_id values contain distinct reviews. Further investigation is required before selecting a single review per order for order-level analysis. Also, multiple reviews may be associated with the same order. Inspection showed both exact duplicate reviews and distinct reviews submitted at different times, suggesting that some customers may have submitted updated reviews. Row order does not consistently reflect review chronology. This will need to be addressed before order-level analysis.
3. There are missing values in the `review_comment_title` and `review_comment_message` columns. For now I will not alter the table and see if it affects the analysis downstream and then deal with it.

In [ ]:
## Loading table No.9 - geolocation
geolocation = pd.read_csv("olist_geolocation_dataset.csv")

## Looking into the raw data
geolocation.head()
print(geolocation.columns)

## Looking for missing values
print("Missing values:\n", geolocation.isnull().sum())

Index(['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng',
       'geolocation_city', 'geolocation_state'],
      dtype='object')
Missing values:
 geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64


### Geolocation table conclusions
In this table each row represents a geolocation record associated with a ZIP code prefix. Multiple records may exist for the same ZIP code prefix.
1. **Primary key:** No natural primary key was identified in the raw dataset.
2. **Interesting points:** This table is a reference table providing geographic coordinates and location information that can be linked to both customers and sellers via ZIP code prefix.
3. There are no missing values in this table.

## Data Model Summary

### Customers
- **Grain:** One row per customer record associated with an order.
- **Primary Key:** `customer_id`
- **Customer Identifier:** `customer_unique_id` identifies the same customer across multiple orders.
- **Connects to:** `orders` via `customer_id`
- **Geographic reference:** `customer_zip_code_prefix` can be linked to `geolocation`.

### Orders
- **Grain:** One row per order.
- **Primary Key:** `order_id`
- **Foreign Key:** `customer_id`
- **Connects to:** `customers`, `order_items`, `order_payments`, and `order_reviews`.

### Order Items
- **Grain:** One row per item within an order.
- **Composite Key:** `order_id` + `order_item_id`
- **Foreign Keys:** `order_id`, `product_id`, `seller_id`
- **Connects to:** `orders`, `products`, and `sellers`.

### Products
- **Grain:** One row per product.
- **Primary Key:** `product_id`
- **Category reference:** `product_category_name`
- **Connects to:** `order_items` via `product_id` and `product_category_translation` via `product_category_name`.

### Product Category Translation
- **Grain:** One row per product category.
- **Primary Key:** `product_category_name`
- **Connects to:** `products` via `product_category_name`.

### Sellers
- **Grain:** One row per seller.
- **Primary Key:** `seller_id`
- **Connects to:** `order_items` via `seller_id`.
- **Geographic reference:** `seller_zip_code_prefix` can be linked to `geolocation`.

### Order Payments
- **Grain:** One payment component within an order.
- **Composite Key:** `order_id` + `payment_sequential`
- **Foreign Key:** `order_id`
- **Connects to:** `orders`.

### Order Reviews
- **Grain:** One review–order association.
- **Composite Key:** `review_id` + `order_id`
- **Foreign Key:** `order_id`
- **Connects to:** `orders`.
- **Important note:** Multiple reviews may be associated with the same order, and the same review record may appear across multiple orders. This will require further handling before order-level analysis.

### Geolocation
- **Grain:** One geolocation record associated with a ZIP code prefix.
- **Primary Key:** No natural primary key identified in the raw dataset.
- **Reference Key:** `geolocation_zip_code_prefix`
- **Connects to:** `customers` and `sellers` through their respective ZIP code prefix fields.


# Phase 2: Data Quality & Cleaning

## Orders - Data Quality & Cleaning

### Data Types

The timestamp columns were initially loaded as `object` and need to be converted to datetime for time-based analysis.

In [ ]:
## Checking the data type in the orders table
print(orders.info())

## Converting 5 date columns from an object data type to datetime data type
date_columns = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date'
]

for col in date_columns:
  orders[col] = pd.to_datetime(orders[col])

## Validating data type
print(orders[date_columns].dtypes)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB
None
order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
dtype: object


### Date and Time Conversion

All order timestamp columns were initially loaded as `object` data types.
They were converted to `datetime` to enable time-based calculations and delivery performance analysis.

### Missing values

In [ ]:
## Looking into the missing values again
print(orders.isnull().sum())

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64


#### Missing Customer Delivery Dates

In [ ]:
# Looking at the values in the order status
print(orders['order_status'].value_counts())

# Out of all the null customer delivered date, what is the order status?
print(orders[
    orders['order_delivered_customer_date'].isnull()
]['order_status'].value_counts())

# 8 records have a delivered status but no delivery date
delivered_missing_date = orders[
    (orders['order_status'] == 'delivered') &
    (orders['order_delivered_customer_date'].isnull())
]

delivered_missing_date

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64
order_status
shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: count, dtype: int64


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,ec05a6d8558c6455f0cbbd8a420ad34f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaT,2017-12-18
20618,f5dd62b788049ad9fc0526e3ad11a097,5e89028e024b381dc84a13a3570decb4,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaT,2018-07-16
43834,2ebdfc4f15f23b91474edf87475f108e,29f0540231702fda0cfdee0a310f11aa,delivered,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaT,2018-07-30
79263,e69f75a717d64fc5ecdfae42b2e8e086,cfda40ca8dd0a5d486a9635b611b398a,delivered,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaT,2018-07-30
82868,0d3268bad9b086af767785e3f0fc0133,4f1d63d35fb7c8999853b2699f5c7649,delivered,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaT,2018-07-24
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaT,NaT,2017-06-23
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,dd1b84a7286eb4524d52af4256c0ba24,delivered,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaT,2018-06-26
98038,20edc82cf5400ce95e1afacc25798b31,28c37425f1127d887d7337f284080a0f,delivered,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaT,2018-07-19


#### Missing Customer Delivery Dates

Most missing values in `order_delivered_customer_date` occur for orders whose status indicates that delivery has not yet been completed (e.g., shipped, canceled, unavailable, processing). These are considered expected missing values rather than data quality issues.

However, 8 orders marked as `delivered` are missing an actual customer delivery timestamp:

- 7 orders contain all previous order timestamps, including `order_delivered_carrier_date`, but are missing the final customer delivery date.
- 1 order is missing both the carrier delivery date and the customer delivery date.

These missing timestamps cannot be reliably inferred from the available data. Therefore, no values were imputed. The records will remain in the dataset but cannot contribute to analyses that require an actual delivery timestamp.

#### Missing Carrier Delivery Dates

In [ ]:
# Out of all the null carrier delivered date, what is the order status?
print(orders[
    orders['order_delivered_carrier_date'].isnull()
]['order_status'].value_counts())

# 2 records have a delivered status but no carrier delivered date
delivered_missing_carrier = orders[
    (orders['order_status'] == 'delivered') &
    (orders['order_delivered_carrier_date'].isnull())
]

delivered_missing_carrier

order_status
unavailable    609
canceled       550
invoiced       314
processing     301
created          5
approved         2
delivered        2
Name: count, dtype: int64


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
73222,2aa91108853cecb43c84a5dc5b277475,afeb16c7f46396c0ed54acb45ccaaa40,delivered,2017-09-29 08:52:58,2017-09-29 09:07:16,NaT,2017-11-20 19:44:47,2017-11-14
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaT,NaT,2017-06-23


#### Missing Carrier Delivery Dates

Most missing `order_delivered_carrier_date` values occur in orders whose status indicates that they had not yet reached the shipping stage, and are therefore considered expected missing values.

Two orders marked as `delivered` are missing the carrier delivery timestamp. One of these orders also lacks the customer delivery timestamp, while the other contains a valid customer delivery date.

No timestamps were imputed. These records cannot contribute to analyses that specifically require the carrier delivery timestamp.

#### Missing order approved at dates

In [ ]:
# Out of all the null order approved dates, what is the order status?
print(orders[
    orders['order_approved_at'].isnull()
]['order_status'].value_counts())

# 14 records have a delivered status but no approved date
delivered_missing_approve = orders[
    (orders['order_status'] == 'delivered') &
    (orders['order_approved_at'].isnull())
]

delivered_missing_approve

order_status
canceled     141
delivered     14
created        5
Name: count, dtype: int64


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
5323,e04abd8149ef81b95221e88f6ed9ab6a,2127dc6603ac33544953ef05ec155771,delivered,2017-02-18 14:40:00,NaT,2017-02-23 12:04:47,2017-03-01 13:25:33,2017-03-17
16567,8a9adc69528e1001fc68dd0aaebbb54a,4c1ccc74e00993733742a3c786dc3c1f,delivered,2017-02-18 12:45:31,NaT,2017-02-23 09:01:52,2017-03-02 10:05:06,2017-03-21
19031,7013bcfc1c97fe719a7b5e05e61c12db,2941af76d38100e0f8740a374f1a5dc3,delivered,2017-02-18 13:29:47,NaT,2017-02-22 16:25:25,2017-03-01 08:07:38,2017-03-17
22663,5cf925b116421afa85ee25e99b4c34fb,29c35fc91fc13fb5073c8f30505d860d,delivered,2017-02-18 16:48:35,NaT,2017-02-22 11:23:10,2017-03-09 07:28:47,2017-03-31
23156,12a95a3c06dbaec84bcfb0e2da5d228a,1e101e0daffaddce8159d25a8e53f2b2,delivered,2017-02-17 13:05:55,NaT,2017-02-22 11:23:11,2017-03-02 11:09:19,2017-03-20
26800,c1d4211b3dae76144deccd6c74144a88,684cb238dc5b5d6366244e0e0776b450,delivered,2017-01-19 12:48:08,NaT,2017-01-25 14:56:50,2017-01-30 18:16:01,2017-03-01
38290,d69e5d356402adc8cf17e08b5033acfb,68d081753ad4fe22fc4d410a9eb1ca01,delivered,2017-02-19 01:28:47,NaT,2017-02-23 03:11:48,2017-03-02 03:41:58,2017-03-27
39334,d77031d6a3c8a52f019764e68f211c69,0bf35cac6cc7327065da879e2d90fae8,delivered,2017-02-18 11:04:19,NaT,2017-02-23 07:23:36,2017-03-02 16:15:23,2017-03-22
48401,7002a78c79c519ac54022d4f8a65e6e8,d5de688c321096d15508faae67a27051,delivered,2017-01-19 22:26:59,NaT,2017-01-27 11:08:05,2017-02-06 14:22:19,2017-03-16
61743,2eecb0d85f281280f79fa00f9cec1a95,a3d3c38e58b9d2dfb9207cab690b6310,delivered,2017-02-17 17:21:55,NaT,2017-02-22 11:42:51,2017-03-03 12:16:03,2017-03-20


#### Missing Order Approval Dates

Of the 160 orders missing an `order_approved_at` timestamp:

- 141 were canceled, which is consistent with orders that may not have reached the approval stage.
- 5 remain in `created` status, making the missing approval timestamp consistent with their current order stage.
- 14 are marked as `delivered` and contain subsequent order timestamps, indicating incomplete approval timestamp data.

No approval dates were imputed, as the actual approval time cannot be reliably inferred. These records will remain in the dataset but cannot contribute to analyses that specifically require an approval timestamp.

### Time Consistency

In [ ]:
## Approval
approval_before_purchase = (orders['order_approved_at'] < orders['order_purchase_timestamp']).sum()
print("Approval before purchase: ", approval_before_purchase)

## Carrier
carrier_before_approval = (orders['order_delivered_carrier_date'] < orders['order_approved_at']).sum()
carrier_before_purchase = (orders['order_delivered_carrier_date'] < orders['order_purchase_timestamp']).sum()
print("Carrier before approval: ", carrier_before_approval)
print("Carrier before purchase: ", carrier_before_purchase)

## Customer delivery
customer_before_carrier = (orders['order_delivered_customer_date'] < orders['order_delivered_carrier_date']).sum()
customer_before_approval = (orders['order_delivered_customer_date'] < orders['order_approved_at']).sum()
customer_before_purchase = (orders['order_estimated_delivery_date'] < orders['order_purchase_timestamp']).sum()
print("Customer before carrier: ", customer_before_carrier)
print("Customer before approval: ", customer_before_approval)
print("Customer before purchase: ", customer_before_purchase)



Approval before purchase:  0
Carrier before approval:  1359
Carrier before purchase:  166
Customer before carrier:  23
Customer before approval:  61
Customer before purchase:  0


We can see that there are some inconsistencies that I want to further investigate. To check whether the time discrepancies are a matter of seconds or days for example.

#### Carrier delivery before purchase

In [ ]:
## Inconsistency between the time of the delivery to the carrier and the purchase time
carrier_before_purchase_rows = orders[
    orders['order_delivered_carrier_date'] <
    orders['order_purchase_timestamp']
].copy()

# Finding the time difference
carrier_before_purchase_rows['time_difference'] = (
    carrier_before_purchase_rows['order_purchase_timestamp'] -
    carrier_before_purchase_rows['order_delivered_carrier_date']
)

# Time difference distribution
carrier_before_purchase_rows['time_difference'].describe()

# Dividing the time differences to 3 groups to better understand the distributions
print(
    "Up to 1 hour:",
    (carrier_before_purchase_rows['time_difference'] <= pd.Timedelta(hours=1)).sum()
)

print(
    "1–24 hours:",
    (
        (carrier_before_purchase_rows['time_difference'] > pd.Timedelta(hours=1)) &
        (carrier_before_purchase_rows['time_difference'] <= pd.Timedelta(days=1))
    ).sum()
)

print(
    "More than 1 day:",
    (carrier_before_purchase_rows['time_difference'] > pd.Timedelta(days=1)).sum()
)

# Looking into the records with the largest time difference
carrier_before_purchase_rows.sort_values(
    'time_difference',
    ascending=False
)[[
    'order_id',
    'order_purchase_timestamp',
    'order_delivered_carrier_date',
    'time_difference'
]].head(10)

Up to 1 hour: 121
1–24 hours: 43
More than 1 day: 2


,order_id,order_purchase_timestamp,order_delivered_carrier_date,time_difference
25883,7c48bb55e8e4f7e56d412e9653db37bc,2018-07-16 18:40:53,2018-01-26 13:35:00,171 days 05:05:53
83321,4021cd7611d6d9ce5ffcd24817fc374f,2018-08-18 11:49:40,2018-08-14 06:22:00,4 days 05:27:40
67844,db090a16182b263b1e896bb26c6f66cf,2018-07-13 16:14:08,2018-07-13 13:59:00,0 days 02:15:08
65452,9711d975b961355b4b5d636857e48498,2018-06-13 15:23:50,2018-06-13 13:15:00,0 days 02:08:50
79401,89d32b64af005178b318f76cd60f2c3c,2018-07-06 11:54:40,2018-07-06 09:48:00,0 days 02:06:40
13969,6192897f85cb2aff6dd1fca56fddb45e,2018-07-18 13:34:22,2018-07-18 11:33:00,0 days 02:01:22
43657,3ed559dec69ab96d9074fedaf18650eb,2018-06-15 13:20:09,2018-06-15 11:22:00,0 days 01:58:09
4256,4e157a36ea9cf89bde6fff57a780b525,2018-08-24 14:37:50,2018-08-24 12:43:00,0 days 01:54:50
74503,119060af27b04f5fae8b2b0e27eef7f9,2018-06-11 14:34:47,2018-06-11 12:45:00,0 days 01:49:47
21910,e23c9dcc304042fa90d538be679fa68d,2018-06-13 11:41:23,2018-06-13 09:53:00,0 days 01:48:23


#### Carrier delivery before approval

In [ ]:
## Inconsistency between the time of the delivery to the carrier and the approval time
carrier_before_approval_rows = orders[
    orders['order_delivered_carrier_date'] <
    orders['order_approved_at']
].copy()

# Finding the time difference
carrier_before_approval_rows['time_difference'] = (
    carrier_before_approval_rows['order_approved_at'] -
    carrier_before_approval_rows['order_delivered_carrier_date']
)

# Time difference distribution
carrier_before_approval_rows['time_difference'].describe()

# checking whether the 180 days anomality is of the same record
carrier_before_approval_rows.sort_values(
    'time_difference',
    ascending=False
)[[
    'order_id',
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'time_difference'
]].head()

,order_id,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,time_difference
25883,7c48bb55e8e4f7e56d412e9653db37bc,2018-07-16 18:40:53,2018-07-16 18:50:22,2018-01-26 13:35:00,171 days 05:15:22
14562,1fab4ac9d85079b3da72a11475ae1685,2017-09-01 19:04:22,2017-09-13 22:06:11,2017-09-04 13:10:23,9 days 08:55:48
46163,0184d4ddb259e1a4cfc2871888cf97b8,2017-09-01 20:04:28,2017-09-13 22:17:15,2017-09-04 14:05:50,9 days 08:11:25
98710,1378f9601350615613cc8832d6789c5d,2017-09-01 20:28:02,2017-09-13 22:03:51,2017-09-04 18:07:55,9 days 03:55:56
41592,8554cb37f7158cb0b082a841d24a4589,2017-09-01 18:40:44,2017-09-13 21:58:04,2017-09-04 19:12:19,9 days 02:45:45


#### Customer delivery before carrier

In [ ]:
## Inconsistency between the time of the delivery to the customer and the time of the delivery to the carrier
customer_before_carrier_rows = orders[
    orders['order_delivered_customer_date'] <
    orders['order_delivered_carrier_date']
].copy()

# Finding the time difference
customer_before_carrier_rows['time_difference'] = (
    customer_before_carrier_rows['order_delivered_carrier_date'] -
    customer_before_carrier_rows['order_delivered_customer_date']
)

# Time difference distribution
customer_before_carrier_rows['time_difference'].describe()

,time_difference
count,23
mean,3 days 06:27:18.478260869
std,3 days 17:18:29.050924326
min,0 days 00:23:18
25%,0 days 23:20:18
50%,1 days 15:51:52
75%,5 days 08:45:32.500000
max,16 days 02:18:29


#### Customer delivery before approval

In [ ]:
## Inconsistency between the time of the delivery to the customer and the approval time
customer_before_approval_rows = orders[
    orders['order_delivered_customer_date'] <
    orders['order_approved_at']
].copy()

# Finding the time difference
customer_before_approval_rows['time_difference'] = (
    customer_before_approval_rows['order_approved_at'] -
    customer_before_approval_rows['order_delivered_customer_date']
)

# Time difference distribution
customer_before_approval_rows['time_difference'].describe()


,time_difference
count,61
mean,1 days 15:03:15.868852459
std,1 days 18:58:33.660093032
min,0 days 00:05:19
25%,0 days 03:27:08
50%,1 days 00:49:24
75%,2 days 07:49:37
max,6 days 23:45:06


### Order Timeline Validation

The chronological consistency of the order lifecycle timestamps was validated by comparing each recorded event with all preceding stages.

Expected sequence:
Purchase → Approval → Carrier → Customer Delivery

The following inconsistencies were identified:

- **Approval before Purchase:** 0 records
- **Carrier before Purchase:** 166 records
- **Carrier before Approval:** 1,359 records
- **Customer Delivery before Purchase:** 0 records
- **Customer Delivery before Approval:** 61 records
- **Customer Delivery before Carrier:** 23 records

Further investigation showed that the 166 Carrier-before-Purchase cases were mostly small time differences: 121 occurred within one hour and 43 within 24 hours. Only two exceeded one day, including one extreme anomaly of approximately 180 days.

The Carrier-before-Approval inconsistencies showed a median difference of approximately 17 hours, with the same 180-day record appearing as the largest anomaly.

Customer-before-Carrier inconsistencies had a median difference of approximately 1 day and 16 hours, while Customer-before-Approval inconsistencies had a median difference of approximately 1 day.

The underlying cause of these chronological inconsistencies cannot be determined from the available data. Therefore, the original timestamps were retained rather than corrected or imputed.

For time-duration analyses, records will only be included when the timestamps required for the specific calculation are present and chronologically valid.

### Consistency in the delivery status

In [ ]:
## Looking to see that the different status of the orders are written in a consistent way
print(orders['order_status'].value_counts())

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64


### Orders Cleaning Summary

- Converted all order timestamp columns from `object` to `datetime`.
- Investigated missing timestamps and distinguished expected missing values based on order status from unexpected missing records.
- Missing timestamps were retained rather than imputed when their true values could not be reliably inferred.
- Validated the chronological consistency of the order lifecycle and identified several timestamp inconsistencies.
- Original timestamps were retained. Original timestamps were retained, and the identified missing or chronologically inconsistent values were documented as data-quality limitations.
- Validated duplicate records and categorical consistency of order statuses.

In [1]:
# Looking at the dirst and last purchase timestamp
orders['order_purchase_timestamp'].min()
orders['order_purchase_timestamp'].max()
orders.groupby('order_status')['order_purchase_timestamp'].agg(['min', 'max'])


NameError: name 'orders' is not defined

## Customers — Data Quality & Cleaning

### Data Types

In [ ]:
## Checking the data type in the customers table
print(customers.info())

## Changing the datatype of the zip code to a string
customers['customer_zip_code_prefix'] = customers['customer_zip_code_prefix'].astype('string')

## Validating the datatype of the zipcode
print(customers['customer_zip_code_prefix'].dtype)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  string
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: object(4), string(1)
memory usage: 3.8+ MB
None
string


### Data Types

`customer_zip_code_prefix` was loaded as an integer. Since ZIP codes function as categorical identifiers rather than numerical measures, the column was converted to a string data type.

### Consistency in the state column

In [ ]:
# State consistency
print(sorted(customers['customer_state'].unique()))

['AC', 'AL', 'AM', 'AP', 'BA', 'CE', 'DF', 'ES', 'GO', 'MA', 'MG', 'MS', 'MT', 'PA', 'PB', 'PE', 'PI', 'PR', 'RJ', 'RN', 'RO', 'RR', 'RS', 'SC', 'SE', 'SP', 'TO']


### Spaces in the city and state names

In [ ]:
# Stripping any unnecessary spaces from the cities and the states
customers['customer_city'] = customers['customer_city'].str.strip()
customers['customer_state'] = customers['customer_state'].str.strip()

### Customers Cleaning Summary

- No missing values were identified in the dataset.
- `customer_id` was previously validated as the primary key, with one unique value per row.
- `customer_unique_id` was retained as the identifier used to recognize the same customer across multiple orders.
- `customer_zip_code_prefix` was converted from an integer to a string, as ZIP codes function as categorical identifiers rather than numerical measures.
- Basic text normalization was applied to `customer_city` and `customer_state` by removing leading and trailing whitespace.
- `customer_state` values were reviewed for categorical consistency and no inconsistencies were identified.

No records required removal or imputation.

## Sellers — Data Quality & Cleaning

### Data Types

In [ ]:
## Checking the data type in the sellers table
print(sellers.info())

## Changing the zip code datatype to string
sellers['seller_zip_code_prefix'] = sellers['seller_zip_code_prefix'].astype('string')

## Validating the new datatype
print(sellers['seller_zip_code_prefix'].dtype)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   string
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: object(3), string(1)
memory usage: 96.8+ KB
None
string


### Data Types

`seller_zip_code_prefix` was loaded as an integer. Since ZIP codes function as categorical identifiers rather than numerical measures, the column was converted to a string data type.

### Consistency in the state column

In [ ]:
# State consistency
print(sorted(sellers['seller_state'].unique()))

['AC', 'AM', 'BA', 'CE', 'DF', 'ES', 'GO', 'MA', 'MG', 'MS', 'MT', 'PA', 'PB', 'PE', 'PI', 'PR', 'RJ', 'RN', 'RO', 'RS', 'SC', 'SE', 'SP']


### Spaces in the city and state names

In [ ]:
# Stripping any unnecessary spaces from the cities and the states
sellers['seller_city'] = sellers['seller_city'].str.strip()
sellers['seller_state'] = sellers['seller_state'].str.strip()

### Sellers Cleaning Summary

- No missing values were identified in the dataset.
- `seller_id` was previously validated as the primary key, with one unique value per row.
- `seller_zip_code_prefix` was converted from an integer to a string, as ZIP codes function as categorical identifiers rather than numerical measures.
- Basic text normalization was applied to `seller_city` and `seller_state` by removing leading and trailing whitespace.
- `seller_state` values were reviewed for categorical consistency and no inconsistencies were identified.

No records required removal or imputation.

## Translation — Data Quality & Cleaning

### Data Types

In [ ]:
## Checking the data type in the translation table
print(translation.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB
None


### Trimming spaces

In [ ]:
translation['product_category_name_english'] = translation['product_category_name_english'].str.strip()
translation['product_category_name'] = translation['product_category_name'].str.strip()

### Product Category Translation Cleaning Summary

- Both category columns were retained as text fields.
- No missing values were identified.
- `product_category_name` was previously validated as unique across all 71 records.
- Leading and trailing whitespace was removed from both category columns to ensure consistent text formatting.

No records required removal or imputation.

## Payments — Data Quality & Cleaning

### Data Types

In [ ]:
## Checking the data type in the payments table
print(payments.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB
None


### Payment type

In [ ]:
## Looking at all the differernt type of payments
print(payments['payment_type'].value_counts())

## 3 are "not defined" and need further investigation
payments[payments['payment_type'] == 'not_defined']
orders[orders['order_id'].isin(
    ['4637ca194b6387e2d538dc89b124b0ee',
     '00b1cb0320190ca0daa2c88b35206009',
     'c8c528189310eaa44a745b8d9d26908b'])
]

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
1130,00b1cb0320190ca0daa2c88b35206009,3532ba38a3fd242259a514ac2b6ae6b6,canceled,2018-08-28 15:26:39,NaN,NaN,NaN,2018-09-12 00:00:00
39919,4637ca194b6387e2d538dc89b124b0ee,a73c1f73f5772cf801434bf984b0b1a7,canceled,2018-09-03 14:14:25,NaN,NaN,NaN,2018-09-10 00:00:00
40235,c8c528189310eaa44a745b8d9d26908b,197a2a6a77da93f678ea0d379f21da0a,canceled,2018-08-28 20:05:14,NaN,NaN,NaN,2018-09-11 00:00:00


#### Undefined Payment Types

Three records have `payment_type = 'not_defined'`. All three have a payment value of 0 and correspond to canceled orders with no approval, carrier, or customer delivery timestamps.

These records were retained without modification, as no actual payment type can be inferred from the available data.

#### Installments, values and sequential

In [ ]:
## Finding out the statistics of the installments
payments['payment_installments'].describe()

## looking into the differernt values of the payment installments
payments['payment_installments'].value_counts().sort_index() # 2 records with 0 installments

## Looking at the entire records of the 0 installments
payments[payments['payment_installments'] == 0]

## Looking at the actual orders with 0 installments
zero_installment_order_ids = payments.loc[
    payments['payment_installments'] == 0,
    'order_id'
]

payments[
    payments['order_id'].isin(zero_installment_order_ids)
].sort_values(['order_id', 'payment_sequential'])



Negative values: 0
Zero values: 9


,order_id,payment_sequential,payment_type,payment_installments,payment_value
33781,45ed6e85398a87c253db47c2d9f48216,1,voucher,1,21.13
11755,45ed6e85398a87c253db47c2d9f48216,2,voucher,1,50.01
62674,45ed6e85398a87c253db47c2d9f48216,3,voucher,1,0.00
40546,6ccb433e00daae1283ccc956189c82ae,1,credit_card,5,84.67
93478,6ccb433e00daae1283ccc956189c82ae,2,voucher,1,14.65
92318,6ccb433e00daae1283ccc956189c82ae,3,voucher,1,22.72
43744,6ccb433e00daae1283ccc956189c82ae,4,voucher,1,0.00
20963,8bcbe01d44d147f901cd3192671144db,1,credit_card,1,36.21
63762,8bcbe01d44d147f901cd3192671144db,2,voucher,1,31.70
5163,8bcbe01d44d147f901cd3192671144db,3,voucher,1,6.25


Two credit card payment records have `payment_installments = 0` despite having positive payment values. Since the correct number of installments cannot be inferred from the available data, these values were retained without modification.

In [ ]:
## Finding out the statistics of the sequentials
payments['payment_sequential'].describe()

## Looking into the different values of payment sequentials
payments['payment_sequential'].value_counts().sort_index()

## Finding out the min number of sequential in each order and grouping it
print(payments.groupby('order_id')['payment_sequential'].min().value_counts())


payment_sequential
1    99360
2       80
Name: count, dtype: int64
Negative values: 0
Zero values: 9


,order_id,payment_sequential,payment_type,payment_installments,payment_value
33781,45ed6e85398a87c253db47c2d9f48216,1,voucher,1,21.13
11755,45ed6e85398a87c253db47c2d9f48216,2,voucher,1,50.01
62674,45ed6e85398a87c253db47c2d9f48216,3,voucher,1,0.00
40546,6ccb433e00daae1283ccc956189c82ae,1,credit_card,5,84.67
93478,6ccb433e00daae1283ccc956189c82ae,2,voucher,1,14.65
92318,6ccb433e00daae1283ccc956189c82ae,3,voucher,1,22.72
43744,6ccb433e00daae1283ccc956189c82ae,4,voucher,1,0.00
20963,8bcbe01d44d147f901cd3192671144db,1,credit_card,1,36.21
63762,8bcbe01d44d147f901cd3192671144db,2,voucher,1,31.70
5163,8bcbe01d44d147f901cd3192671144db,3,voucher,1,6.25


Out of all the orders, for 80 orders the minimum value of sequential payment is 2 and the rest is 1.

In [ ]:
## Finding out the statistics of the payment values
payments['payment_value'].describe()

## Printing the suspicious values
print("Negative values:", (payments['payment_value'] < 0).sum())
print("Zero values:", (payments['payment_value'] == 0).sum())

## Looking at orders with payment value of 0.
zero_payment_rows = payments[
    payments['payment_value'] == 0
]

zero_payment_rows # Apart from the 3 not-defined payment method, the other 6 are of voucher type

## Looking at the orders that have a voucher payment of 0
zero_voucher_order_ids = payments.loc[
    (payments['payment_value'] == 0) &
    (payments['payment_type'] == 'voucher'),
    'order_id'
]

payments[
    payments['order_id'].isin(zero_voucher_order_ids)
].sort_values(['order_id', 'payment_sequential'])

Negative values: 0
Zero values: 9


,order_id,payment_sequential,payment_type,payment_installments,payment_value
33781,45ed6e85398a87c253db47c2d9f48216,1,voucher,1,21.13
11755,45ed6e85398a87c253db47c2d9f48216,2,voucher,1,50.01
62674,45ed6e85398a87c253db47c2d9f48216,3,voucher,1,0.00
40546,6ccb433e00daae1283ccc956189c82ae,1,credit_card,5,84.67
93478,6ccb433e00daae1283ccc956189c82ae,2,voucher,1,14.65
92318,6ccb433e00daae1283ccc956189c82ae,3,voucher,1,22.72
43744,6ccb433e00daae1283ccc956189c82ae,4,voucher,1,0.00
20963,8bcbe01d44d147f901cd3192671144db,1,credit_card,1,36.21
63762,8bcbe01d44d147f901cd3192671144db,2,voucher,1,31.70
5163,8bcbe01d44d147f901cd3192671144db,3,voucher,1,6.25


#### Zero Payment Values

Nine payment records have a `payment_value` of 0.

- Three have an undefined payment type and correspond to canceled orders with no recorded payment activity.
- Six are voucher records belonging to orders that also contain other positive-value payments.

No negative payment values were identified. The zero-value records were retained, as their context does not provide sufficient justification for modification or removal.

### Payments Cleaning Summary

- All column data types were appropriate and required no conversion.
- No missing values were identified.
- `order_id` and `payment_sequential` were previously validated as the composite key.
- Three `not_defined` payment types were investigated and found to correspond to canceled orders with zero payment value.
- Two credit card records contain `payment_installments = 0` despite having positive payment values. Since the correct values cannot be inferred, they were retained without modification.
- Payment installment values otherwise ranged from 1 to 24, with no additional validity concerns identified.
- No negative payment values were found.
- Nine zero-value payment records were investigated: three correspond to the canceled `not_defined` records, while six are voucher records in orders containing other positive-value payments.
- No values were imputed or records removed.

## Products — Data Quality & Cleaning

### Data Types

In [ ]:
## Checking the data type in the products table
print(products.info())

## Converting the name_length, description_length and photos_qty to Int64 datatype
integer_columns = [
    'product_name_lenght',
    'product_description_lenght',
    'product_photos_qty'
]

products[integer_columns] = products[integer_columns].astype('Int64')

## Validating datatype
print(products.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  Int64  
 3   product_description_lenght  32341 non-null  Int64  
 4   product_photos_qty          32341 non-null  Int64  
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: Int64(3), float64(4), object(2)
memory usage: 2.4+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      

### Missing values
We already know that we have 610 records that have missing values. We will see if these products are also part of the actual orders.

In [ ]:
## New variable of the IDs of the products with NaN values
missing_product_ids = products.loc[
    products['product_category_name'].isna(),
    'product_id'
]

## How many of these products were actually sold
missing_product_sales = order_items[order_items['product_id'].isin(missing_product_ids)]
print("Order items:", len(missing_product_sales))

## Out of the sold products, how many have NaN values
print("Unique products sold:", missing_product_sales['product_id'].nunique())

## The total sales value of these products
print("Total sales value:", missing_product_sales['price'].sum())

Order items: 1603
Unique products sold: 610
Total sales value: 179535.28


### Missing Values

Missing values were identified in several product attributes.

#### Product Information

610 products are missing values for:
- `product_category_name`
- `product_name_lenght`
- `product_description_lenght`
- `product_photos_qty`

To determine whether these products were relevant to the sales analysis, their `product_id` values were compared with the `order_items` table.

All 610 products appear in `order_items`, accounting for 1,603 order-item records in total. Therefore, although these products represent only approximately 1.85% of the products table, they were all involved in actual orders.

The records were retained because the missing attributes do not make the remaining product and sales information unusable. No missing values were imputed, as their true values cannot be reliably inferred.

Products with missing attributes cannot contribute directly to analyses requiring those attributes. For example, products with a missing category cannot be directly included in category-level analysis, while products with missing photo counts cannot be included in analyses involving the number of product photos.

In [ ]:
## Looking into the statisctics of the dimension columns
products[[
    'product_weight_g',
    'product_length_cm',
    'product_height_cm',
    'product_width_cm'
]].describe()

## 4 records were found to have a weight of 0
products[products['product_weight_g'] == 0]

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
9769,81781c0fed9fe1ad6e8c81fca1e1cb08,cama_mesa_banho,51,529,1,0.0,30.0,25.0,30.0
13683,8038040ee2a71048d4bdbbdc985b69ab,cama_mesa_banho,48,528,1,0.0,30.0,25.0,30.0
14997,36ba42dd187055e1fbe943b2d11430ca,cama_mesa_banho,53,528,1,0.0,30.0,25.0,30.0
32079,e673e90efa65a5409ff4196c038bb5af,cama_mesa_banho,53,528,1,0.0,30.0,25.0,30.0


#### Product Measurement Validation

Product weight and dimensions were reviewed for invalid or implausible values.

No negative values were identified, and the length, height, and width measurements showed no immediate validity concerns.

Four products have a recorded weight of 0 grams despite having non-zero physical dimensions. Since a physical product cannot realistically have zero weight and the correct values cannot be inferred from the available data, these records were retained without modification.

The missing product weight values were retained in the dataset. Analyses involving product weight therefore rely only on records with available weight data.

In [ ]:
## Trimming leading and trailing spaces
products['product_category_name'] = products['product_category_name'].str.strip()

In [ ]:
## Looking at the statistics of the following columns
products[[
    'product_name_lenght',
    'product_description_lenght',
    'product_photos_qty'
]].describe()

,product_name_lenght,product_description_lenght,product_photos_qty
count,32341.0,32341.0,32341.0
mean,48.476949,771.495285,2.188986
std,10.245741,635.115225,1.736766
min,5.0,4.0,1.0
25%,42.0,339.0,1.0
50%,51.0,595.0,1.0
75%,57.0,972.0,3.0
max,76.0,3992.0,20.0


### Products Cleaning Summary

- `product_id` was previously validated as the primary key, with one unique value per row.
- Product name length, description length, and photo quantity were converted to nullable integer (`Int64`) data types, as these variables represent discrete counts.
- Product weight and dimensions were retained as numerical continuous variables.
- Leading and trailing whitespace was removed from `product_category_name`.

No records were removed and no missing or invalid values were imputed.

## Order items — Data Quality & Cleaning

### Data Types

In [ ]:
## Checking the data types in the order items table
print(order_items.info())

## Changing the shipping limit date to a datetime data type
order_items['shipping_limit_date'] = pd.to_datetime(
    order_items['shipping_limit_date']
)

## Validating data type
print(order_items['shipping_limit_date'].dtypes)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB
None
datetime64[ns]


In [ ]:
## Looking for 0 or negative values in the prices and freight columns
print("Negative prices:", (order_items['price'] < 0).sum())
print("Zero prices:", (order_items['price'] == 0).sum())

print("Negative freight values:", (order_items['freight_value'] < 0).sum())
print("Zero freight values:", (order_items['freight_value'] == 0).sum())

Negative prices: 0
Zero prices: 0
Negative freight values: 0
Zero freight values: 383


#### Price and Freight Validation

- No zero or negative product prices were identified.
- No negative freight values were identified.
- 383 order items have a `freight_value` of 0. Since zero-cost shipping is a plausible business scenario, these records were retained without modification.

#### Timeline inconsistencies

In [ ]:
## Merging the order_item table and the orders table on the order)id column
order_items_check = order_items.merge(
    orders[['order_id', 'order_purchase_timestamp', 'order_approved_at']],
    on='order_id',
    how='left'
)

## Summing the number of occasions where the shipping limit date was either before the purchase or before the approval
print(
    "Shipping deadline before purchase:",
    (order_items_check['shipping_limit_date'] <
     order_items_check['order_purchase_timestamp']).sum()
)

print(
    "Shipping deadline before approval:",
    (order_items_check['shipping_limit_date'] <
     order_items_check['order_approved_at']).sum()
)

## There are 127 records with shipping deadline before approval date
## Chacking the time difference between them
deadline_before_approval = order_items_check[
    order_items_check['shipping_limit_date'] <
    order_items_check['order_approved_at']
].copy()

deadline_before_approval['time_difference'] = (
    deadline_before_approval['order_approved_at'] -
    deadline_before_approval['shipping_limit_date']
)

## The time differernce statistics
deadline_before_approval['time_difference'].describe()

## Dividing into 3 categories of time differences
print(
    "Up to 1 hour:",
    (deadline_before_approval['time_difference'] <= pd.Timedelta(hours=1)).sum()
)

print(
    "1–24 hours:",
    (
        (deadline_before_approval['time_difference'] > pd.Timedelta(hours=1)) &
        (deadline_before_approval['time_difference'] <= pd.Timedelta(days=1))
    ).sum()
)

print(
    "More than 1 day:",
    (deadline_before_approval['time_difference'] > pd.Timedelta(days=1)).sum()
)

## Looking at the values with the time difference in the descending order
deadline_before_approval[
    ['order_id', 'shipping_limit_date', 'order_approved_at', 'time_difference']
].sort_values('time_difference', ascending=False)

Shipping deadline before purchase: 0
Shipping deadline before approval: 127
Up to 1 hour: 7
1–24 hours: 60
More than 1 day: 60


,order_id,shipping_limit_date,order_approved_at,time_difference
101394,e5fa5a7210941f7d56d0208e4e071d35,2016-09-19 00:15:34,2016-10-07 13:17:15,18 days 13:01:41
47304,6b80bb20190715d71c43efff617bd659,2017-02-23 01:15:03,2017-03-01 10:51:46,6 days 09:36:43
655,0184d4ddb259e1a4cfc2871888cf97b8,2017-09-07 22:31:42,2017-09-13 22:17:15,5 days 23:45:33
3069,06eb87385425e5797a1a5c2cdb1b6559,2017-09-07 22:31:16,2017-09-13 22:10:37,5 days 23:39:21
86162,c3b8c17ee15e0e798c2e178b7d4c7f42,2017-09-07 22:31:45,2017-09-13 22:08:55,5 days 23:37:10
...,...,...,...,...
25163,39454163ea6db630f644e723c00f7da3,2018-07-05 15:31:27,2018-07-05 16:05:14,0 days 00:33:47
13876,1f8e863921f1d0daeda98b603a956954,2018-07-05 15:32:09,2018-07-05 16:03:39,0 days 00:31:30
98523,df719e82188cc47552d259703d5413f0,2018-08-01 16:10:33,2018-08-01 16:31:43,0 days 00:21:10
71522,a31e3f43f9ed2f1d94b0748bc723b899,2018-07-05 16:31:29,2018-07-05 16:34:56,0 days 00:03:27


#### Shipping Limit Date Validation

The `shipping_limit_date` was compared with the corresponding purchase and approval timestamps from the `orders` table.

- No order items had a shipping deadline earlier than the order purchase timestamp.
- 127 order items had a shipping deadline earlier than the recorded order approval timestamp.
- Of these 127 records:
  - 7 had a difference of up to 1 hour.
  - 60 had a difference between 1 and 24 hours.
  - 60 had a difference greater than 24 hours.
- The largest observed difference was approximately 18.5 days.
- Several of the larger inconsistencies occurred around similar dates, suggesting a recurring pattern rather than an isolated record.

The source of these chronological inconsistencies cannot be determined from the available data. Therefore, no timestamps were modified and no records were removed. These records should be treated with caution in analyses that depend specifically on the relationship between order approval and the seller shipping deadline.

## Geolocation — Data Quality & Cleaning

### Data Types

In [ ]:
## Looking into the data types of the geolocation table
print(geolocation.info())

## Changing the zip-code data type to string
geolocation['geolocation_zip_code_prefix'] = geolocation['geolocation_zip_code_prefix'].astype('string')

## Validating data type
print(geolocation['geolocation_zip_code_prefix'].dtypes)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  string 
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), object(2), string(1)
memory usage: 38.2+ MB
None
string


In [ ]:
## Trimming leading and lagging spaces in the city and state columns
geolocation['geolocation_city'] = geolocation['geolocation_city'].str.strip()
geolocation['geolocation_state'] = geolocation['geolocation_state'].str.strip()

In [ ]:
## Looking for consistency in the state names
geolocation['geolocation_state'].value_counts().sort_index()

,count
geolocation_state,
AC,1301
AL,4183
AM,2432
AP,853
BA,36045
CE,11674
DF,12986
ES,16748
GO,20139


State names were cheched for inconsistent capitalization and none were found.
Leading and lagging spaces were trimmed for both the state and the city columns.

In [ ]:
## Verifying that the coordinates in the table are valid
print(
    "Invalid latitude:",
    ((geolocation['geolocation_lat'] < -90) |
     (geolocation['geolocation_lat'] > 90)).sum()
)

print(
    "Invalid longitude:",
    ((geolocation['geolocation_lng'] < -180) |
     (geolocation['geolocation_lng'] > 180)).sum()
)

Invalid latitude: 0
Invalid longitude: 0


No invalid coordinates were found.

### Geolocation Cleaning Summary

- `geolocation_zip_code_prefix` was converted from integer to string, as it represents a geographic identifier rather than a numerical measure.
- Latitude and longitude were retained as floats and were found to fall within valid coordinate ranges.
- Leading and trailing whitespace was removed from city and state values.
- State values were consistent.
- No additional data modifications were required.

## Reviews — Data Quality & Cleaning

### Data Types

In [ ]:
## Looking into the data types of the reviews table
print(reviews.info())

## Changing the creation_date and the answer_timestamp to datetime data type
reviews['review_answer_timestamp'] = pd.to_datetime(reviews['review_answer_timestamp'])
reviews['review_creation_date'] = pd.to_datetime(reviews['review_creation_date'])

## Validating the change in the data type
print(reviews.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                99224 non-null  object        
 1   order_id                 99224 non-null  object        
 2   review_score             99224 non-null  int64         
 3   review_comment_title     11568 non-null  object        
 4   review_comment_message   40977 non-null  object        
 5   review_creation_date     99224 non-null  datetime64[ns]
 6   review_answer_timestamp  99224 non-null  datetime64[ns]
dtypes: datetime64[ns](2), int64(1), object(4)
memory usage: 5.3+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                99224 non

### Dates inconsistency check

In [ ]:
## Checking that the creation of the review was prior to the answer
(
    reviews['review_answer_timestamp'] <
    reviews['review_creation_date']
).sum()

np.int64(0)

### Handling with duplicates

In [ ]:
## Looking for the duplicates
reviews.duplicated().sum()

np.int64(0)

In [ ]:
## Understanding how many records appear more than once for the review_id and the order_id columns
review_id_counts = reviews['review_id'].value_counts()
order_id_counts = reviews['order_id'].value_counts()

print(
    "Review IDs appearing more than once:",
    (review_id_counts > 1).sum()
)

print(
    "Order IDs appearing more than once:",
    (order_id_counts > 1).sum()
)

Review IDs appearing more than once: 789
Order IDs appearing more than once: 547


### Review ID and Order ID Validation

Neither `review_id` nor `order_id` is fully unique in the reviews dataset.

- 789 `review_id` values appear more than once.
- 547 `order_id` values appear more than once.
- No fully duplicated rows were identified.
- Inspection of repeated `order_id` values showed that multiple review records associated with the same order may contain different review content, scores, or timestamps.
- Repeated `review_id` values may also be associated with different `order_id` values.

Since the available data does not provide sufficient information to determine which records, if any, should be considered incorrect, no records were removed or modified.

The non-unique relationship between orders and reviews will be taken into account during analysis and data modeling, particularly for metrics calculated at the order level.

## Final Data Cleaning Summary

The nine datasets were reviewed for data types, missing values, duplicate records, categorical consistency, numerical validity, and relevant chronological inconsistencies. Relationships between tables were also considered where necessary to evaluate the impact of data quality issues.

### Data Type and Formatting Adjustments

- Date and timestamp columns were converted from `object` to `datetime` where required.
- Identifier fields that do not represent numerical quantities, such as ZIP code prefixes, were treated as strings.
- Product name length, description length, and photo quantity were converted to nullable integer (`Int64`) types.
- Leading and trailing whitespace was removed from relevant categorical text fields.

### Missing Values

Missing values were evaluated based on their business meaning rather than being automatically removed or imputed.

- Missing order timestamps were largely consistent with the corresponding order status and therefore represent expected missingness for orders that had not reached those stages.
- A small number of delivered orders were missing timestamps from earlier stages of the order lifecycle. These records were retained because the missing values could not be reliably inferred.
- 610 products are missing category, name-length, description-length, and photo-quantity information. All 610 products appear in `order_items`, accounting for 1603 order-item records, so they were retained.
- Two of these products are also missing physical dimensions and weight.
- Missing review titles and comments were retained, since customers can provide a numerical review score without written feedback.

No missing values were imputed where the true value could not be reliably inferred.

### Numerical and Categorical Validation

- Product prices, freight values, payment values, product dimensions, and other numerical attributes were reviewed for unusual or invalid values.
- Zero freight values were retained because zero-cost shipping is a plausible business scenario.
- Zero-value payment records were investigated and retained based on their payment and order context.
- Two credit-card payment records contain zero recorded installments despite positive payment values. Since the correct installment values cannot be inferred, they were retained.
- Four products have a recorded weight of 0 grams despite having non-zero dimensions. These values were retained because their correct weights could not be reliably inferred.
- Review scores contained only the expected values from 1 to 5.
- Geographic coordinates were within valid latitude and longitude ranges.
- Relevant categorical fields were checked for formatting and consistency issues.

### Chronological Validation

Order lifecycle timestamps were reviewed for chronological inconsistencies.

Several inconsistencies were identified between recorded order stages, including cases where carrier or delivery timestamps occurred before earlier recorded stages. The magnitude of these differences was investigated rather than assuming that all inconsistencies represented data-entry errors.

Additionally, 127 order-item records have a seller shipping deadline earlier than the recorded order approval timestamp.

Since the available data does not provide enough information to determine the source or correct value of these timestamp inconsistencies, the records were retained without modification. They should be treated carefully in analyses that depend on the affected timestamp relationships.

### Duplicate and Key Validation

Primary and composite keys were evaluated during the initial data-understanding stage.

- No unnecessary full-row duplicates requiring removal were identified during cleaning.
- The reviews table contains non-unique relationships between `review_id` and `order_id`: 789 review IDs and 547 order IDs appear more than once.
- Inspection showed that repeated order IDs may correspond to different review records, while repeated review IDs may also be associated with different orders.
- Since there is insufficient information to determine which records, if any, are incorrect, these records were retained.

The non-unique relationship between orders and reviews must be considered when calculating order-level review metrics.

### Final Cleaning Decisions

The cleaning process prioritized preserving valid business information rather than removing records solely because they contained incomplete or unusual values.

No values were artificially imputed when the correct value could not be determined, and records with partial data were retained when their remaining fields were still analytically useful.

Known data-quality limitations were documented and will be handled according to the requirements of each specific analysis.

In [ ]:
## Creating two samples of 25 random reviews with score of 1 and 5
sample_1 = reviews[
    (reviews["review_score"] == 1) &
    (reviews["review_comment_message"].notna())
].sample(n=25, random_state=46)

sample_5 = reviews[
    (reviews["review_score"] == 5) &
    (reviews["review_comment_message"].notna())
].sample(n=25, random_state=46)

review_sample = pd.concat([
    sample_1,
    sample_5
])

review_sample = review_sample.sort_values("review_score")
pd.set_option('display.max_colwidth', None)

## Looking into the reviews to draw conclusions from it
review_sample[review_sample['review_score'] == 1]
review_sample[review_sample['review_score'] == 5]

#### Exploratory Review Analysis
A reproducible random sample of 25 one-star and 25 five-star reviews with written comments was examined to identify recurring themes. This qualitative sample is exploratory and is not intended to represent the full review dataset.